README
Homepage:
    http://www.steve.org.uk/Software/slaughter/

Mercurial Repository:
    http://slaughter.repository.steve.org.uk/

GitHub Mirror:
    https://github.com/skx/slaughter/



Slaughter
---------

The goal of this project is to have a lightweight system which will
allow the control and manipulation of multiple systems.

The tasks we support are pretty basic, but they include:

   1.  Replacing a file, literally.
   2.  Replacing a file, template-expanding content within it.
   3.  Appending lines to files, commenting out lines, replacing lines, etc.
   4.  Running system commands.
   5.  Testing for the existence of local users, and fetching their details.
   6.  Checking disk space, and mount-points.
   7.  Sending alerts via email.

The implementation is cleanly organized to allow it to be ported, tested,
and used upon new systems easily.  Currently we support:

   * GNU/Linux.
   * OpenBSD.
   * Microsoft Windows - using Strawberry perl.



Overview
--------

We assume that we have a "server" somewhere.  This server is designed to
allow files, modules, and policies to be retrieved from it.

The actual form of the server loosely specified, as slaughter allows different
transport mechanisms to be used.  Currently there are several supported transports
for fetching things from the remote server:

   * HTTP-transfers.
   * rsync-transfers.
   * Cloning remote git/mercurial/svn repositories.

Each client which is running slaughter will connect to the server, using
the appropriate mechanism, and download policies.  These policies describe
the actions that must be carried out upon the local system.

Policies (and also files) are pulled by the client, meaning there is no
central server which is in charge of initiating transactions, or connections.

   i.e. Slaughter is client-pull rather than server-push.

The attraction of client-pull is that there is no need to maintain state
on the central server, and each client can be trusted to schedule itself
via a cron-like system.  The potential downside is that you might fail to
notice if a client suddenly stops making appropriate requests.



Policies
--------

The list of tasks which each node should carry out is defined in a policy
file.  An initial "default.policy" file is fetched from the server, and
this may trigger the fetch of additional policies and modules in turn.

This policy files may contain code in (standard) Perl, and they may also
use the primitives that Slaughter provides.

It is expected that because the file default.policy is fetched by all clients
it should be used for house-keeping, and merely include other policies.

For example a default.policy file might look like this:

---

# actions to carry out globally
FetchPolicy "global.policy";

# is there a per-client one?
FetchPolicy "$fqdn.policy";

----

In this case the variable "fqdn" is expanded to the fully-qualified
domain-name of the requesting client - this is an example of one of the
many available defined variables clients may make use of in policy
files, or pure perl code.

If the hostname-specific policy file does not exist then it is merely
ignored.



Client Layout
-------------

To get started a client needs to have :

  a.  The slaughter-client package installed upon it.
  b.  The name of the server, and the transport to use against it, stored
     in the configuration file /etc/slaughter/slaughter.conf [*]

Once this is done cron may be used to ensure that /sbin/slaughter is invoked
upon a regular basis.  (Hourly is a good choice.)

You'll probably want to invoke slaughter manually for the first few times
as you're putting together your policies.

  [*] - Alternatively you may specify these details on the command line.
        See the TRANSPORT file for some examples.



Server Layout
-------------

Regardless of the transport which is in use, and examples are available in
the included TRANSPORT file, slaughter does insist that the top-level of
the server contain the some fixed subdirectories.

For example using the HTTP-transport the webserver should be configured to
serve a tree which would look like this:

           /var/www/slaughter/
           /var/www/slaughter/modules/
           /var/www/slaughter/policies/
           /var/www/slaughter/files/

Here we see there are two specific directories:

  modules/
    This is the location of any modules which are implemented.
  policies/
    This is the location of the policies.
  files/
    This is the root of any files which are stored on the server to be fetched.

e.g. The request for http://$master/slaughter/policies/default.policy should
succeed.

In the interests of security it is probably wise to limit access to the
/slaughter/ location, denying access to clients you're not expecting to
pull from it.




In Depth Operation
------------------

The client node will first fetch the policy, which might contain
references to other policies to load.

Once the policy fetching has been completed the downloaded
content will be wrapped such that it becomes a locally executable file,
making use of the Slaughter.pm module (this module is where the primitives
are implemented and exported to the perl script).

Finally the locally written script will be executed, before being removed.

After your script(s) have executed any messages saved with the
LogMessage function will be displayed, or sent via email.

(Any literal output your script might have written with "print" will
be sent to STDOUT, as expected.)


Steve
--


debian/changelog
slaughter2 (2.5-1) stable; urgency=low

  * The git/rsync/svn/etc transports are quiet by default.
  * Module loading allows more than a single module to be loaded.

 -- Steve Kemp <steve@steve.org.uk>  Fri, 08 Feb 2013 11:33:11 +0000

slaughter2 (2.4-1) stable; urgency=low

  * New release.  Features module-handling.

 -- Steve Kemp <steve@steve.org.uk>  Thu, 07 Feb 2013 14:00:41 +0000

slaughter2 (2.3-1) stable; urgency=low

  * Cleaned-up and centralized the expansion of policy fetches.

 -- Steve Kemp <steve@steve.org.uk>  Thu, 01 Jan 2013 21:21:21 +0000

slaughter2 (2.2-1) stable; urgency=low

  * Added new primitive "IdenticalContents" to compare file bodies.
  * Updated dependencies so that we can build upon squeeze hosts,
    that just means added "libparent-perl".

 -- Steve Kemp <steve@steve.org.uk>  Wed, 26 Dec 2012 05:44:50 +0000

slaughter2 (2.1-1) stable; urgency=low

  * Updated to extend the information available to the clients
    This now includes "memtotal", "memfree", and CPU counts.

 -- Steve Kemp <steve@steve.org.uk>  Tue, 18 Dec 2012 08:33:08 +0000

slaughter2 (2.0-1) stable; urgency=low

  * Updated to use a new system for fetching files/policies, using
    pluggable transports.  Available transports include:
    - local
    - http
    - hg
    - git
    - rsync
    - svn
  * Renamed the client package to slaughter2 to avoid breakage on upgrade,
    but also update the configuration file post-install, to mitigate against this.
  * The information module determines IPv6 address details on Linux clients.
  * Updated to ensure that polices are unreadable to non-root users.
  * Updated correctness of several primitives.
  * Updated documentation: 100% POD coverage.
  * Added test-suite.
  * Implemented and tested our primitives and package-API on OpenBSD.
  * Implemented and tested our primitives and package-API on FreeBSD.
  * New primitive LogMessage, which can be sent by email.
  * New primitive UserCreate.

 -- Steve Kemp <steve@steve.org.uk>  Mon, 03 Dec 2012 08:20:08 +0000

slaughter (1.7-1) stable; urgency=low

  * BUGFIX Release:
    - SHA1 sum comparisons succeed as expected.

 -- Steve Kemp <steve@steve.org.uk>  Tue, 09 Oct 2012 18:27:18 +0000

slaughter (1.6-1) stable; urgency=low

  * New release containing integrated test suite.
  * New primitive SetPermissions.
  * Bugfixes.
  * SSL validation of https:// based fetches & slaughter-specific user-agent.

 -- Steve Kemp <steve@steve.org.uk>  Tue, 09 Oct 2012 08:30:08 +0000

slaughter (1.5-1) stable; urgency=low

  * Abstract away the meta-information collection into a series of modules.
  * Renamed the Linux package-helping library.
  * Updated to use more obvious logfile names.
  * Added support for Microsoft Windows clients, via Strawberry Perl.
  * Allow template expansion in policy filenames.
  * Included manpage for "slaughter".
  * New command line flag: --version

 -- Steve Kemp <steve@steve.org.uk>  Sat, 26 Nov 2011 14:44:41 +0000

slaughter (1.4-1) stable; urgency=low

  * Updated to fetch $URL.$ARCH for each file fetch operation,
    and log the failures of URL fetches generally.
  * Updated documentation, available via "slaughter --manu".
  * Updated standards version.

 -- Steve Kemp <steve@steve.org.uk>  Tue, 1 Nov 2011 20:40:02 +0000

slaughter (1.3-1) stable; urgency=low

  * Updated so that we have '$release' holding the LSB release name
    of the current host.  Useful now I have at least one Squeeze host.

 -- Steve Kemp <steve@steve.org.uk>  Wed, 16 Feb 2011 21:40:12 +0000

slaughter (1.2-1) stable; urgency=low

  * Update the IP-address detection to cope with "ip" being located
    in either /sbin or /bin.

 -- Steve Kemp <steve@steve.org.uk>  Tuesday, 23 Nov 2010 21:11:12 +0000

slaughter (1.1-1) stable; urgency=low

  * Suggestions implemented from Raphael Daum:
    - Expose IP address of client in "ip0, ip1, ip..., ipN".
    - IP count available in "ipcount".
    - Read /etc/hostname if available, and only then revert to
      `hostname` and `hostname --fqdn` to find FQDN.
    - Honour comments when fetching policies.

 -- Steve Kemp <steve@steve.org.uk>  Monday, 22 Nov 2010 20:30:40 +0000

slaughter (1.0-1) stable; urgency=low

  * Fixed file download to correctly compare existing contents.
    - Thanks to Kristian Lyngstol.

 -- Steve Kemp <steve@steve.org.uk>  Sunday, 18 Apr 2010 11:22:11 +0000

slaughter (0.9-2) stable; urgency=low

  * Experimentation.

 -- Steve Kemp <steve@steve.org.uk>  Monday, 25 Jan 2010 23:55:32 +0000

slaughter (0.9-1) stable; urgency=low

  * New release:
    - Added new --no-log parameter to avoid local logging.
    - Changed DeleteFilesMatching to return the number of files deleted.

 -- Steve Kemp <steve@steve.org.uk>  Sunday, 24 Jan 2010 12:04:04 +0000

slaughter (0.8-1) stable; urgency=low

  * New release:
    - Added "DeleteOldFiles".
    - Added more verbosity in some of the primitives.

 -- Steve Kemp <steve@steve.org.uk>  Monday, 18 Jan 2010 03:48:59 +0000

slaughter (0.7-1) stable; urgency=low

  * New release:
    - Removes:  Define, Defined, and Value.  Which are redundent.
    - Install HACKING as a documentation file.

 -- Steve Kemp <steve@steve.org.uk>  Monday, 18 Jan 2010 00:05:00 +0000

slaughter (0.6-1) stable; urgency=low

  * New release as a clean package.

 -- Steve Kemp <steve@steve.org.uk>  Sunday, 17 Jan 2010 22:22:22 +0000

slaughter (0.5-1) stable; urgency=low

  * Package abstraction is here.  We can now install on at least my
    Debian sid machine and my Centos 4.7 install.

 -- Steve Kemp <steve@steve.org.uk>  Sunday, 17 Jan 2010 20:00:02 +0000

slaughter (0.4-4) stable; urgency=low

  * More minor cleanup.

 -- Steve Kemp <steve@steve.org.uk>  Sunday, 17 Jan 2010 17:16:18 +0000

slaughter (0.4-3) stable; urgency=low

  * Minor cleanup.

 -- Steve Kemp <steve@steve.org.uk>  Sunday, 17 Jan 2010 05:22:50 +0000

slaughter (0.4-2) stable; urgency=low

  * Added "PackageInstalled".  TODO decouple the debian packaging

lib/Slaughter/Info/linux.pm
#!/usr/bin/perl -w

=head1 NAME

Slaughter::Info::linux - Perl Automation Tool Helper linux info implementation

=cut

=head1 SYNOPSIS

This module is the GNU/Linux version of the Slaughter information-gathering
module.

Modules beneath the C<Slaughter::Info> namespace are loaded when slaughter
is executed, they are used to populate a hash with information about
the current host.

This module is loaded only on linux systems, and will determine such details
as the local hostname, the free RAM, any IP addresses, etc.

Usage is:

=for example begin

    use Slaughter::Info::linux;

    my $obj  = Slaughter::Info::linux->new();
    my $data = $obj->getInformation();

    # use info now ..
    print "We have software RAID\n" if ( $data->{'softwareraid'} );

=for example end

When this module is used an attempt is also made to load the module
C<Slaughter::Info::Local::linux> - if that succeeds it will be used to
augment the information discovered and made available to slaughter
policies.

=cut


=head1 AUTHOR

 Steve
 --
 http://www.steve.org.uk/

=cut

=head1 LICENSE

Copyright (c) 2010-2012 by Steve Kemp.  All rights reserved.

This module is free software;
you can redistribute it and/or modify it under
the same terms as Perl itself.
The LICENSE file contains the full text of the license.

=cut


use strict;
use warnings;


package Slaughter::Info::linux;




=head2 new

Create a new instance of this object.

=cut

sub new
{
    my ( $proto, %supplied ) = (@_);
    my $class = ref($proto) || $proto;

    my $self = {};
    bless( $self, $class );
    return $self;

}


=head2 getInformation

This function retrieves meta-information about the current host.

The return value is a hash-reference of data determined dynamically.

=cut

sub getInformation
{
    my ($self) = (@_);

    #
    #  The data we will return.
    #
    my $ref;

    #
    #  Fully Qualified hostname
    #
    #  1.  If we can find /etc/hostname, then use that.
    #
    if ( -e "/etc/hostname" )
    {
        open( my $file, "<", "/etc/hostname" ) or
          die "Failed to read /etc/hostname - $!";
        $ref->{ 'fqdn' } = <$file>;
        chomp( $ref->{ 'fqdn' } );
        close($file);
    }
    else
    {

        #
        #  Call "hostname".
        #
        $ref->{ 'fqdn' } = `hostname`;
        chomp( $ref->{ 'fqdn' } );

        #
        # If it is unqualified retry with --fqdn.
        #
        if ( $ref->{ 'fqdn' } !~ /\./ )
        {
            $ref->{ 'fqdn' } = `hostname --fqdn`;
            chomp( $ref->{ 'fqdn' } );
        }
    }


    #
    #  Get the hostname and domain name as seperate strings.
    #
    if ( $ref->{ 'fqdn' } =~ /^([^.]+)\.(.*)$/ )
    {
        $ref->{ 'hostname' } = $1;
        $ref->{ 'domain' }   = $2;
    }
    else
    {

        #
        #  Better than nothing, right?
        #
        $ref->{ 'hostname' } = $ref->{ 'fqdn' };
        $ref->{ 'domain' }   = $ref->{ 'fqdn' };
    }

    #
    # This should be portable.
    #
    $ref->{ 'path' } = $ENV{ 'PATH' };

    #
    #  Is this a xen host, or guest?
    #
    $ref->{ 'xen' } = 1 if -d "/proc/xen/capabilities";

    #
    #  Detect virtualized CPU, as well as processor counts.
    #
    if ( open( my $cpu, "<", "/proc/cpuinfo" ) )
    {
        $ref->{ 'cpu_count' } = -1;

        foreach my $line (<$cpu>)
        {
            chomp($line);
            $ref->{ 'kvm' } = 1 if ( $line =~ /model/ && $line =~ /qemu/i );

            if ( $line =~ /model name\s+: (.*)$/ )
            {
                $ref->{ 'cpumodel' } = $1;
            }
            if ( $line =~ /processor\s+: (\d+)/ )
            {
                $ref->{ 'cpu_count' } = $1 if ( $ref->{ 'cpu_count' } < $1 );
            }
        }

        $ref->{ 'cpu_count' }++;
        close($cpu);
    }


    #
    #  Software RAID?
    #
    if ( ( -e "/proc/mdstat" ) &&
         ( -x "/sbin/mdadm" ) )
    {
        if ( open( my $mdstat, "<", "/proc/mdstat" ) )
        {
            foreach my $line (<$mdstat>)
            {
                if ( ( $line =~ /^md([0-9]+)/ ) &&
                     ( $line =~ /active/i ) )
                {
                    $ref->{ 'softwareraid' } = 1;
                    $ref->{ 'raid' }         = "software";
                }
            }
            close($mdstat);
        }
    }


    #
    #  Memory total and memory free.
    #
    if ( open( my $mem, "<", "/proc/meminfo" ) )
    {
        foreach my $line (<$mem>)
        {
            chomp($line);
            if ( $line =~ /MemTotal:\s+(\d+) kB/ )
            {
                $ref->{ 'memtotal' } = $1;
            }
            if ( $line =~ /MemFree:\s+(\d+) kB/ )
            {
                $ref->{ 'memfree' } = $1;
            }
        }
        close($mem);
    }


    #
    #  Kernel version.
    #
    $ref->{ 'kernel' } = `uname -r`;
    chomp( $ref->{ 'kernel' } );

    #
    #  Are we i386/amd64?
    #
    my $type = `file /bin/ls`;
    if ( $type =~ /64-bit/i )
    {
        $ref->{ 'arch' } = "amd64";
        $ref->{ 'bits' } = 64;
    }
    else
    {
        $ref->{ 'arch' } = "i386";
        $ref->{ 'bits' } = 32;
    }

    #
    #  IP address(es).
    #
    my $ip = undef;

    $ip = "/sbin/ip" if ( -x "/sbin/ip" );
    $ip = "/bin/ip"  if ( -x "/bin/ip" );


    if ( defined($ip) )
    {

        #
        #  Two commands to find the IP addresses we have
        #
        my @cmd = ( " -o -f inet addr show scope global",
                    " -o -f inet6 addr show scope global"
                  );

        #
        #  Run each
        #
        foreach my $cmd (@cmd)
        {
            my $count  = 1;
            my $family = "ip";
            $family = "ip6" if ( $cmd =~ /inet6/i );

            foreach my $line ( split( /[\r\n]/, `$ip $cmd` ) )
            {
                next if ( !defined($line) || !length($line) );
                chomp($line);

                #
                #  This matches something like:
                #
                #  eth0 inet 192.168.1.9/24 brd 192.168.1.255 scope global eth0
                #
                # or
                #  eth0 inet6 2001:41c8:1:5abb::62/64 scope global valid_lft forever preferred_lft forever
                #
                #
                if ( $line =~ /(inet|inet6)[ \t]+([^ \t+]+)[ \t]+/ )
                {
                    my $proto = $1;
                    my $ip    = $2;

                    #
                    #  Strip off /24, /128, etc.
                    #
                    $ip =~ s/\/.*//g;

                    #
                    # Save away the IP address in "ip0", "ip1", "ip2" .. etc.
                    #
                    $ref->{ $family . "_" . $count } = $ip;
                    $count += 1;
                }
            }

            if ( $count > 0 )
            {
                $ref->{ $family . '_count' } = ( $count - 1 );
            }
        }
    }

    #
    #  Find the name of our release
    #
    my $version = "unknown";
    my $distrib = "unknown";
    my $release = "unknown";
    if ( -x "/usr/bin/lsb_release" )
    {
        foreach
          my $line ( split( /[\r\n]/, `/usr/bin/lsb_release -a 2>/dev/null` ) )
        {
            chomp $line;
            if ( $line =~ /Distributor ID:\s*(.*)/ )
            {
                $distrib = $1;
            }
            if ( $line =~ /Release:\s*(.*)/ )
            {
                $version = $1;
            }
            if ( $line =~ /Codename:\s*(.*)/ )
            {
                $release = $1;
            }
        }
    }
    $ref->{ 'version' }      = $version;
    $ref->{ 'distribution' } = $distrib;
    $ref->{ 'release' }      = $release;


    #
    #  TODO: 3Ware RAID?
    #

    #
    #  TODO: HP RAID?
    #

    #
    # Load Average
    #
    my $uptime = `uptime`;
    chomp($uptime);
    if ( $uptime =~ /load average:[ \t]*(.*)/ )
    {
        $uptime = $1;
        $uptime =~ s/,//g;
        $ref->{ 'load_average' } = $uptime;

        #
        #  Split into per-minute values.
        #
        my @avg = split( /[ \t]/, $ref->{ 'load_average' } );
        $ref->{ 'load_average_1' }  = $avg[0];
        $ref->{ 'load_average_5' }  = $avg[1];
        $ref->{ 'load_average_15' } = $avg[2];

    }


    return ($ref);
}



1;
